rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,221 @@
+    let pos = unsafe { *pos };
+    match T::start(arg) {
+        Some(mut wrapper) => {
+            for _ in 0..pos {
+                wrapper.as_mut().next();
+            }
+            Box::into_raw(wrapper) as *mut c_types::c_void
+        }
+        None => ptr::null_mut(),
+    }
+}
+
+struct SeqFileOperationsVTable<T>(PhantomData<T>);
+
+impl<T: SeqOperations> SeqFileOperationsVTable<T> {
+    const VTABLE: bindings::seq_operations = bindings::seq_operations {
+        start: Some(start_callback::<T>),
+        stop: Some(stop_callback::<T>),
+        next: Some(next_callback::<T>),
+        show: Some(show_callback::<T>),
+    };
+}
+
+/// A `seq_file` referencing data of type `T`.
+pub struct SeqFile<T> {
+    seq_ops: bindings::seq_operations,
+    _pin: PhantomPinned,
+    dir_entry: *mut bindings::proc_dir_entry,
+    context: T,
+}
+
+impl<T> Drop for SeqFile<T> {
+    fn drop(&mut self) {
+        if !self.dir_entry.is_null() {
+            // SAFETY: Calling a C function. `dir_entry` is a valid pointer to a
+            // `proc_dir_entry` because if it is not null it was created by a
+            // call to `bindings::proc_create_seq_private`.
+            unsafe {
+                bindings::proc_remove(self.dir_entry);
+            }
+        }
+    }
+}
+
+impl<T> SeqFile<T> {
+    fn new(seq_ops: bindings::seq_operations, context: T) -> KernelResult<Box<Self>> {
+        Ok(Box::try_new(SeqFile {
+            seq_ops,
+            _pin: PhantomPinned,
+            dir_entry: ptr::null_mut::<bindings::proc_dir_entry>(),
+            context,
+        })?)
+    }
+
+    /// Retrieve the [`SeqFile::context`] associated with a [`bindings::seq_file`].
+    ///
+    /// # Safety
+    ///
+    /// `m` must have been created from a proc file generated by calling
+    /// `proc_create::<T>`.
+    unsafe fn convert(m: *mut bindings::seq_file) -> *const T {
+        let reg = crate::container_of!((*m).op, Self, seq_ops);
+        &(*reg).context
+    }
+}
+
+// SAFETY: The fields of `SeqFile` that are not `Sync` are `dir_entry` and
+// potentially `context`, but `SeqFile` provides no access to these fields.
+unsafe impl<T> Sync for SeqFile<T> {}
+
+/// Create a `/proc` file.
+///
+/// The returned value must not be dropped until the module is unloaded.
+pub fn proc_create<T: SeqOperations + Sync>(
+    name: CStr<'static>,
+    context: T,
+) -> KernelResult<Pin<Box<SeqFile<T>>>> {
+    let mut reg = SeqFile::new(SeqFileOperationsVTable::<T>::VTABLE, context)?;
+
+    let name = name.deref().as_ptr() as *const u8 as *const c_types::c_char;
+
+    // SAFETY: Calling a C function. `name` is guaranteed to be null terminated because it is of type
+    // `CStr`.
+    let dir_entry = unsafe {
+        bindings::proc_create_seq_private(
+            name,
+            0,
+            ptr::null_mut(),
+            &reg.seq_ops,
+            0,
+            ptr::null_mut(),
+        )
+    };
```

Comment:
Shouldn't we fail `proc_create` if `proc_create_seq_private` fails?

This would also simplify `SeqFile` slightly as `dir_entry` will always be known to be non-null.

Reply:
Makes sense, although I'm not sure which error code to use. Switched it to `EINVAL` for now.

Reply:
Saw that `debugfs` returns `ENOMEM` when failing to create a `dentry`, so switched to using that here.